csrc/multidevice/utils.h

```diff
@@ -45,4 +44,16 @@ void unshard(TensorView*);
 // TODO: add an option to rather insert the Set AFTER the resharding Expr
 void insertReshardings(Fusion* fusion);
 
+// This can only run after the insertResharding pass.
+// Assumes all resharding ops are either a set or reduction.
+// For each resharding operation that requires communication
+// over a noncontiguous slices of the tensor, this pass
+// inserts permutations necessary to push the device parallel axis
+// to the front so that communication operations are contiguous.
+void insertShardedAxisReordering(Fusion* fusion);
+
+// Returns the index of the a sharded axis if none return -1.
+// TODO: Assumes no merges/splits on sharded axis.
+int64_t getShardedAxis(TensorView*);
```

Comment:
```suggestion
// Returns the index of the first sharded axis or -1 if none.
// TODO: Assumes no merges/splits on sharded axis.
int64_t getFirstShardedAxis(TensorView*);
```

Reply:
Nit: I believe we will support multiple sharded axes in the not too distant future. When that happens, it's useful to remember by name this function returns merely the first sharded axis.

Reply:
If I'm correct we should only accept one axis assigned to one parallel type, i.e., two axis of a given tensor cannot be both parallelized on DIDx, but rather one can be parallelized DIDx and the other on DIDy. I think there is the same reasonable asumption for TID, BID.
So I think the comment shouldn't mention "first", but rather "DIDx", and later it could take the parallel type as an argument